File: Java/Medium/Accepted/shopping-offers_2.java
class Solution {
    public int shoppingOffers(List<Integer> price, List<List<Integer>> special, List<Integer> needs) {
        
        return shopping(price, special, needs, 0);
    }
    
    private int shopping(List<Integer> price, List<List<Integer>> special, List<Integer> needs, int start) {
        int res = directPurchase(price, needs);
        for (int i = start; i < special.size(); i++) {
            List<Integer> offer = special.get(i);
            List<Integer> clone = new ArrayList<>(needs);
            int j = 0;
            for (; j < needs.size(); j++) {
                int diff = clone.get(j) - offer.get(j);
                if (diff < 0) {
                    break;
                }
                clone.set(j, diff);
            }
            if (j == needs.size()) {
                res = Math.min(res, offer.get(j) + shopping(price, special, clone, i));
            }
        }
        return res;
    }
    
    private int directPurchase(List<Integer> price, List<Integer> needs) {
        int res = 0;
        for (int i = 0; i < price.size(); i++) {
            res += price.get(i) * needs.get(i);
        }
        return res;
    }
}

File: Python3/Medium/TimeLimitExceeded/shopping-offers_2.py
class Solution:
    def shoppingOffers(self, price: List[int], special: List[List[int]], needs: List[int]) -> int:
        
        
        def dfs(needs):
            res = sum([price[i]*needs[i] for i in range(len(needs))])
            for offer in special:
                new_needs = [needs[i]-offer[i] for i in range(len(needs))]
                if min(new_needs) >= 0:
                    res = min(res, offer[-1]+dfs(new_needs))
            return res
        
        return dfs(needs)